app/scripts/controllers/permissions/background-api.js

```diff
@@ -1,38 +1,90 @@
 import nanoid from 'nanoid';
+import { MethodNames } from '@metamask/permission-controller';
+import {
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+} from '../../lib/multichain-api/caip25permissions';
+import {
+  getEthAccounts,
+  setEthAccounts,
+} from '../../lib/multichain-api/adapters/caip-permission-adapter-eth-accounts';
+import { setPermittedEthChainIds } from '../../lib/multichain-api/adapters/caip-permission-adapter-permittedChains';
 import {
   CaveatTypes,
   RestrictedMethods,
 } from '../../../../shared/constants/permissions';
-import { CaveatFactories } from './specifications';
+import { PermissionNames } from './specifications';
 
-export function getPermissionBackgroundApiMethods(permissionController) {
+export function getPermissionBackgroundApiMethods({
+  permissionController,
+  approvalController,
+  networkController,
+}) {
+  // To add more than one account when already connected to the dapp
   const addMoreAccounts = (origin, accountOrAccounts) => {
     const accounts = Array.isArray(accountOrAccounts)
       ? accountOrAccounts
       : [accountOrAccounts];
-    const caveat = CaveatFactories.restrictReturnedAccounts(accounts);
-
-    permissionController.grantPermissionsIncremental({
-      subject: { origin },
-      approvedPermissions: {
-        [RestrictedMethods.eth_accounts]: { caveats: [caveat] },
-      },
-    });
+
+    let caip25Caveat;
+    try {
+      caip25Caveat = permissionController.getCaveat(
+        origin,
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+      );
+    } catch (err) {
+      // noop
+    }
+
+    if (!caip25Caveat) {
+      throw new Error('tried to add accounts when none have been permissioned'); // TODO: better error
```

Comment:
Was an error thrown upstream in this case before? Or is this just an improvement in logging?

Reply:
Yes, originally an error would have been thrown by the `permissionController.grantPermissionsIncremental()` call if the permission did not exist. We have to perform this check ourselves now.

I'm not sure if this needs a better error though. This API is internal facing

Reply:
I think this is okay for now since its internal!